packages/hash/src.ts/ens-normalize/lib.ts

```diff
@@ -27,112 +27,638 @@
+			/*
+			if (CM.has(decomposed[i])) {
+				let j = i + 1;
+				while (j < e && CM.has(decomposed[j])) j++;
+				if (j - i > M) {
+					throw new Error(`too many combining marks: ${g.N} ${bidi_qq(str_from_cps(decomposed.slice(i-1, j)))} (${j-i}/${M})`);
+				}
+				i = j;
+			}
+			*/
+			// 20230217: switch to NSM counting
+			// https://www.unicode.org/reports/tr39/#Optional_Detection
+			if (decomposed[i] in NSM) {
+				let j = i + 1;
+				for (let cp; j < e && (cp = decomposed[j]) in NSM; j++) {
+					// a. Forbid sequences of the same nonspacing mark.
+					for (let k = i; k < j; k++) { // O(n^2) but n < 100
+						if (decomposed[k] == cp) {
+							throw new Error(`duplicate non-spacing marks: ${quoted_cp(cp)}`);
+						}
+					}
+				}
+				// parse to end so we have full nsm count
+				// b. Forbid sequences of more than 4 nonspacing marks (gc=Mn or gc=Me).
+				if (j - i > NSM_MAX) {
+					// note: this slice starts with a base char or spacing-mark cm
+					throw new Error(`excessive non-spacing marks: ${bidi_qq(safe_str_from_cps(decomposed.slice(i-1, j)))} (${j-i}/${NSM_MAX})`);
+				}
+				i = j;
+			}
+		}
+	}
+	// *** this code currently isn't needed ***
+	/*
+	let cm_whitelist = M instanceof Map;
+	for (let i = 0, e = cps.length; i < e; ) {
+		let cp = cps[i++];
+		let seqs = cm_whitelist && M.get(cp);
+		if (seqs) { 
+			// list of codepoints that can follow
+			// if this exists, this will always be 1+
+			let j = i;
+			while (j < e && CM.has(cps[j])) j++;
+			let cms = cps.slice(i, j);
+			let match = seqs.find(seq => !compare_arrays(seq, cms));
+			if (!match) throw new Error(`disallowed combining mark sequence: "${safe_str_from_cps([cp, ...cms])}"`);
+			i = j;
+		} else if (!V.has(cp)) {
+			// https://www.unicode.org/reports/tr39/#mixed_script_confusables
+			let quoted = quoted_cp(cp);
+			for (let cp of cps) {
+				let u = UNIQUE.get(cp);
+				if (u && u !== g) {
+					// if both scripts are restricted this error is confusing
+					// because we don't differentiate RestrictedA from RestrictedB 
+					if (!u.R) quoted = `${quoted} is ${u.N}`;
+					break;
+				}
+			}
+			throw new Error(`disallowed ${g.N} character: ${quoted}`);
+			//throw new Error(`disallowed character: ${quoted} (expected ${g.N})`);
+			//throw new Error(`${g.N} does not allow: ${quoted}`);
+		}
+	}
+	if (!cm_whitelist) {
+		let decomposed = nfd(cps);
+		for (let i = 1, e = decomposed.length; i < e; i++) { // we know it can't be cm leading
+			if (CM.has(decomposed[i])) {
+				let j = i + 1;
+				while (j < e && CM.has(decomposed[j])) j++;
+				if (j - i > M) {
+					throw new Error(`too many combining marks: "${str_from_cps(decomposed.slice(i-1, j))}" (${j-i}/${M})`);
+				}
+				i = j;
+			}
+		}
+	}
+	*/
+}
+
+// given a list of codepoints
+// returns a list of lists, where emoji are a fully-qualified (as Array subclass)
+// eg. explode_cp("abc💩d") => [[61, 62, 63], Emoji[1F4A9, FE0F], [64]]
```

Comment:
No non-ASCII7 characters are allowed in the source code (applies to v6 as well as v5) as they can cause issues in some editors, over various transports, etc. It also helps prevent "trojan" code.